### DISCUSSION TITLE
montage without masking and preserve alpha

### DISCUSSION THREAD

**Discussion by indexofrefraction** [ImageMagick/ImageMagick#5068] [Help]
hi everybody, 

i'm failing at using montage on a bunch of 64x64 tif icons 
the icons have a grey background (space around the icons) plus an alpha channel.

**montage -mode concatenate -tile 8  icon_*.tif result.tif**
X the result has no alpha channel at all.
X the former alpha channels have been applied as a mask and the space around the icons is now white.

**montage -mode concatenate -tile 8 -background None icon_*.tif result.tif**
X the  alpha channels also have been applied as a mask and the space around the icons is now black
√ the merged alpha channel itself is correct.

**montage -mode concatenate -tile 8 -background None -alpha off icon_*.tif result.tif**
√ the space around the icons is untouched (gray)
X but the merged alpha channel is completely white, the alpha information was cleared. 

i simple need to merge the icons into a tiles file **without any channel changes**. 
after hours of tests i'm posting here to maybe get some help .-)

**Comment by snibgo** [comment]
> the icons have a grey background (space around the icons)

With "-background None":
> and the space around the icons is now black

I don't understand this. Please show some example icon files.

**Comment by indexofrefraction** [comment]
tx for helping! here to explain a merge of three icons on a 2x2 tiles file

this is a sample icon (the color around the icon is grey, there is an alpha channel)
![sample_icon](https://user-images.githubusercontent.com/25868186/164886412-79eb738b-dfc9-4adb-8d25-14e61084f04f.jpg)
this is what i need (adjoined tiles without any channel changes) 
![result_desired](https://user-images.githubusercontent.com/25868186/164886416-ddb9cd6f-ad0c-4273-83be-c08da4f22bfd.jpg)
this is what i get 
the alpha channel is correct, but the rgb channels have been altered (masked)
![result_achieved](https://user-images.githubusercontent.com/25868186/164886421-0c4ea7d6-f433-43e3-a807-57d388923a05.jpg)

additional... 
the resulting tiles file needs X colums and X rows, 
so i guess i need montage -mode contencate

edit...
please notice the **separate** alpha channel in the images above!

**Comment by snibgo** [comment]
Okay, thanks, maybe I can see what you are trying to do.

Please post or link to some actual icon files. Not screengrabs, but the actual files.

**Reply by indexofrefraction** [reply]
there 3 dots to test on ... 
https://www61.zippyshare.com/v/C7gc7ipS/file.html or https://send.cm/d/AMDc

the full command i used is
**montage -mode concatenate -tile 2 -channel rgba -background None -define tiff:alpha=unspecified -compress LZW -colorspace sRGB icon_*.tif result.tif**
here i get the alpha channel but a black background

with only
**montage -mode concatenate -tile 3 -channel rgba -background None icon_*.tif result.tif**
i get a transparent background and no alpha channel

(samples / goal updated above)

**Comment by indexofrefraction** [comment]
even tough i'd prefer to have a single command without temp files
i at least succeeded now in getting the parts

**montage -mode concatenate -tile 2 -alpha off icon_*.tif result_rgb.tif**
correct rgb channels without alpha channel
**montage -mode concatenate -tile 2 -alpha extract icon_*.tif result_alpha.tif**
correct alpha channel without rgb channels

maybe it is possible to merge these
(without alpha masking the rgb channels)

**Comment by snibgo** [comment]
Your zippyshare link says "403 Forbidden".

**Reply by indexofrefraction** [reply]
ah, sorry... i think zippy is blocked in some places, here an alternative
https://send.cm/d/AMDc

**Comment by fmw42** [comment]
Why don't you just use -append or -smush?

https://imagemagick.org/script/command-line-options.php#append
https://imagemagick.org/script/command-line-options.php#smush
https://legacy.imagemagick.org/Usage/layers/#append

**Reply by indexofrefraction** [reply]
tx, but i think these commands only allow to append tiles either horizontally or vertically (?)
in the end i need a fixed number of columns with multiple rows... 
if this is possible with something else than **montage -mode concatenate -tile N** i'd be all ears :-)

**Reply by fmw42** [reply]
You can append in separate rows and then append the rows.

**Reply by indexofrefraction** [reply]
yes, but it doesn't solve the problem.
i think now the sources do not have a real alpha, what shows in PSD as "Alpha 1" is a maybe a separate channel.
magick takes it as alpha and masks the rgb channels so that the sources gray background is gone.
i simply need everything exactly as in the sources without any changes in the channels.

**Reply by fmw42** [reply]
Post your PSD file zipped or a link to it and I will look at it in PSD.  If it has an alpha channel as opposed to background transparency, ImageMagick may not see it. ImageMagick generally reads the background transparency from PSD and puts it into a separate alpha channel in a PNG or TIFF file. Photoshop alpha channels are not generally transparency, but separate channels. There has been some attempt at reading them, but I am not sure it is working properly from other reports.  But see the -define psd:alpha-unblend=off at https://imagemagick.org/script/defines.php

**Reply by indexofrefraction** [reply]
there is no PSD file just tifs... all is in this thread already, PSD is only used to check the channels.
plz, just open the source icons (tifs) from the download link in PSD and check the channels. 
then open a result tif and compare. they are not the same. either there is no "Alpha 1" channel in PSD 
and / or the grey background in the RGB channel has been changed (to transparent or black)
the screenshots of the desired RGB and A(?) channels are posted in post 3.

**Reply by fmw42** [reply]
All 3 of your tiff files are RGBA (RGB with an alpha channel). I do not know what photoshop is doing to the files when you open them, since my disk is too full to use my Photoshop. But I opened them in another tool which says they are normal 32-bit RGBA files. I will leave the rest to snibgo, since he has been helping you more.

**Comment by snibgo** [comment]
```
magick montage -tile 2x2 -background None -geometry 64x64+0+0 icon_0?.tif x.png
```
Does that do what you want? Transparent pixels remain transparent. I can't see any effects from the thumbnail operation.

However, transparent gray (approx 30%) has changed to transparent black. Is that the problem you need to fix? If it is, insert "-compose Copy" before reading the inputs.

Including "-alpha off" will make all pixels opaque, which I think you do not want.

**Reply by indexofrefraction** [reply]
tx snibgo,

but, hm... i don't get what you describe...

**magick montage -tile 2x2 -background None -geometry 64x64+0+0 icon_0?.tif result.tif**
this "applies" the alpha. the background gets tranparent. (the rgb channels are masked)
there is no separate alpha channel in the result. see image below.

**magick montage -tile 2x2 -background None -geometry 64x64+0+0 -compose Copy icon_0?.tif result.tif**
i don't see a difference here to the above.

btw i'm using ImageMagick 7.1.0-30 on macOS

![Screenshot 2022-04-23 at 22 49 14](https://user-images.githubusercontent.com/25868186/164945533-423300ee-0a8e-41fa-9563-1e633255c55b.jpg)

**Comment by indexofrefraction** [comment]
maybe it would be easier to go from my extracted parts

starting from the shared folder :

montage -mode concatenate -tile 2 -alpha off icon_*.tif result_rgb.tif
correct rgb channels without alpha channel
![rgb](https://user-images.githubusercontent.com/25868186/164945733-8aab7c38-e32e-4d26-a6ca-213b98360616.jpg)

montage -mode concatenate -tile 2 -alpha extract icon_*.tif result_alpha.tif
correct alpha channel without rgb channels (grayscale image)
![alpha](https://user-images.githubusercontent.com/25868186/164945734-ce92269b-84be-49af-ab54-e43650eea6be.jpg)

anyone with an idea on how to merge these so that 
- the RGB channels are NOT altered in any way (grey background preserved)
- with the alpha as a **separate channel** like in the 2nd picture in post 3

**Comment by snibgo** [comment]
This is the result of my command "magick montage -tile 2x2 -background None -geometry 64x64+0+0 -compose Copy icon_0?.tif x.png", x.png:

![x](https://user-images.githubusercontent.com/33812368/164946034-e036d354-01f6-4871-a270-e161208812f4.png)
The parts that may look white on your screen are fully transparent. Do you get the same result?

You are doing more work to make you final images, for tab captions "Layers" and "Channels", and so on. What software does that? If ImageMagick, what is your command? I suspect that software is showing the transparent parts as a chequerboard. And it adds a black square border around the image. Is that the black you don't want?

**Reply by indexofrefraction** [reply]
no that is a misunderstanding, i'm not doing more... 
i simply open the result in PSD and go to the channels
and in the case of your result, i see:
- a transparent background around the circles (in the R,G,B channels)
- no separate "Alpha 1" channel

i need unaltered R,G,B channels (the sources grey background preserved) and the alpha as separate 4th channel.

i also thought about this being a software issue, because transparent means there must be an alpha channel somehow.
but nonetheless the sources show a separate "Alpha 1" channel in PSD... the result doesn't.
its like image magick "bakes" the alpha in, which is not what i need.
i just need the sources positioned next to each other, without alpha masking or baking or however to call this.

maybe the "Alpha 1" channel is not really an alpha channel, its just an additional "selection" channel.

**Comment by snibgo** [comment]
> i simply open the result in PSD and go to the channels

What is PSD? I don't know it.

icon_01.tif has four channels. exiftool says the first three are RGB, then:
```
Extra Samples                   : Unspecified
```
exiftool also says:
```
Alpha Channels Names            : Alpha
```
IM treats the fourth channel as alpha, meaning that it records transparency.

My command makes an image with an alpha channel, for transparency.

So, the pictures your are showing are screenshots from "PSD"? With Gimp, we can see a list of channels:
![c](https://user-images.githubusercontent.com/33812368/164947715-c66118c6-7373-42f2-add8-bd7d9226de79.png)

Gimp shows four channels, including alpha. I don't know why "PSD" doesn't show alpha.

As far as I know, IM has no facility for naming channels, for example naming alpha as "Alpha 1". Perhaps exiftool can do that.

**Reply by indexofrefraction** [reply]
PSD = Photoshop

**Reply by indexofrefraction** [reply]
we're getting closer :)

 **magick montage -mode concatenate -tile 2 -background None icon_0?.tif result.tif**
no "Alpha 1" channel in PSD, background of RGB channels shows transparent
 **magick montage -mode concatenate -tile 2 -background None icon_0?.tif -define tiff:alpha=unspecified result.tif**
now i get the "Alpha 1" channel in PSD, but the background of RGB channels show black

maybe it is something about how the channel is defined in the tiff file 
and depending on this PSD shows the alpha as channel or not ...

**Comment by snibgo** [comment]
> ... but the background of RGB channels show black

Try my "-compose Copy" suggestion:

```
magick montage -mode concatenate -tile 2 -background None -compose Copy icon_0?.tif -define tiff:alpha=unspecified result.tif
```

EDIT: also, if needed, turn off the alpha channel (click on the eye symbol).

**Reply by indexofrefraction** [reply]
thanks snibgo,

it was late yesterday and i needed a rest here .-)
i downloaded gimp for macos today and tested your result(s) from above.
the grey background has been altered and is black, so the RGB channels are not the same anymore.
...

**Comment by indexofrefraction** [comment]
thanks guys for helping yesterday! ... 
these are my thoughts right now :

if i open my source "icon_01.tif" in gimp on macos i get a dialog : ! Extra channels with unspecified data.
Process exta channel as "Non-premultiplied alpha", "Premultiplied alpha" or "Channel"

if i choose "Premultiplied alpha" the grey background pixel information is gone.
with the other options the grey background has not been altered, 
i can check this by clicking on the eye symbol of the alpha channel.

in PSD i do not get such a dialog.
I guess depending on the -define tiff:alpha=unspecified option
PSD treats it as an extra channel "Alpha 1" shown in the channel list
or as alpha, which is then (in contrary to Gimp) not shown at all in the channel list

your posts helped me at least to get a better understanding of what is happening.

—> as a conclusion there are serveral factors :
- how the alpha is described in the tif format (alpha, extra channel / mask)
- how the software i verify the files with (PSD, Gimp, etc) interprets the file 
- how magick interprets the file, if it "applies" the alpha and modifies the grey background or not.

obviously magick interprets the unspecified channel as alpha
and when merging images it tends to apply the alpha and  just copies the alpha selected part 
onto a transparent canvas which results in a loss of the original grey background pixels.

i will try to find out now if magick can treat the "alpha" as extra cannnel
or maybe the behaviour is different using append/smush or using -tile 2x2 
this is also something to explore

**Comment by indexofrefraction** [comment]
ok, finally something is working :)

**montage -mode concatenate -tile 2x -alpha off icon_*.tif rgb.tif**
--> correct rgb channels without alpha channel
**montage -mode concatenate -tile 2x -alpha extract icon_*.tif mask.tif**
--> correct alpha channel without rgb channels (grayscale image)
**convert rgb.tif mask.tif -compose copy-opacity -composite -define tiff:alpha=unspecified result.tif**
--> result exactly the same as sources. grey background preserved. alpha not premultiplied. PSD shows the channel.

a single command would be nicer (and faster) ofc, but it will do for now
